core/server/services/members/middleware.js

```diff
@@ -51,18 +51,31 @@ const exchangeTokenForSession = async function (req, res, next) {
     if (!req.url.includes('token=')) {
         return next();
     }
+
+    Object.assign(req, {action: 'signin'});
+
     try {
         const member = await membersService.ssr.exchangeTokenForSession(req, res);
         Object.assign(req, {member});
         next();
     } catch (err) {
         common.logging.warn(err.message);
+        if (err.name === 'TokenExpiredError') {
+            Object.assign(req, {error: {
+                message: 'Your signin link has expired, please send another login link.'
```

Comment:
Don't think its necessary or safe to pass a whole error object to the client. But wanted to stay with the standard shape of the error object till the last stage where the "action" object passed to the HBS is formed